pkg/kube/fakeclient.go

```diff
@@ -0,0 +1,397 @@
+
+	if NewCrdWatcher != nil {
+		c.crdWatcher = NewCrdWatcher(c)
+	}
+
+	return c
+}
+
+func NewFakeClientWithVersion(minor string, objects ...runtime.Object) CLIClient {
+	c := NewFakeClient(objects...).(*client)
+	c.Kube().Discovery().(*fakediscovery.FakeDiscovery).FakedServerVersion = &kubeVersion.Info{Major: "1", Minor: minor, GitVersion: fmt.Sprintf("v1.%v.0", minor)}
+	return c
+}
+
+type fakeClient interface {
+	PrependReactor(verb, resource string, reaction clienttesting.ReactionFunc)
+	PrependWatchReactor(resource string, reaction clienttesting.WatchReactionFunc)
+	Tracker() clienttesting.ObjectTracker
+}
+
+func insertPatchReactor(f clienttesting.FakeClient, fmf *fieldManagerFactory) {
+	f.PrependReactor(
+		"patch",
+		"*",
+		func(action clienttesting.Action) (handled bool, ret runtime.Object, err error) {
+			pa := action.(clienttesting.PatchAction)
+			if pa.GetPatchType() == types.ApplyPatchType {
+				// Apply patches are supposed to upsert, but fake client fails if the object doesn't exist,
+				// if an apply patch occurs for a deployment that doesn't yet exist, create it.
+				// However, we already hold the fakeclient lock, so we can't use the front door.
+				// rfunc := clienttesting.ObjectReaction(f.Tracker())
+				original, err := f.Tracker().Get(pa.GetResource(), pa.GetNamespace(), pa.GetName())
+				var isNew bool
+				mygvk := gvk.MustFromGVR(pa.GetResource()).Kubernetes()
+				if kerrors.IsNotFound(err) || original == nil {
+					isNew = true
+					original = kubeclient.GVRToObject(pa.GetResource())
+					d := original.(metav1.Object)
+					d.SetName(pa.GetName())
+					d.SetNamespace(pa.GetNamespace())
+					e := original.(schema.ObjectKind)
+					e.SetGroupVersionKind(mygvk)
+				}
+
+				applyObj := &unstructured.Unstructured{Object: map[string]interface{}{}}
+				if err := json.Unmarshal(pa.GetPatch(), &applyObj.Object); err != nil {
+					log.Errorf("error decoding YAML: %v", err)
+					return false, nil, err
+				}
+				// TODO: I'm not 100% sure this should always force, but we lose the apply option...
+				res, err := fmf.FieldManager(mygvk).Apply(original, applyObj, "istio", true)
+				if err != nil {
+					log.Errorf("error applying patch: %v", err)
+					return false, res, err
+				}
+
+				// field manager outputs typed objects.  We might need unustructured here.
+				var typedResult runtime.Object
+				if _, ok := f.(*dynamicfake.FakeDynamicClient); ok {
+					typedResult = &unstructured.Unstructured{}
+					err = fmf.schema.Convert(res, typedResult, nil)
+					if err != nil {
+						log.Errorf("error converting object (%v), fakeclients may become inconsistent: %v", res, err)
+						return false, nil, err
+					}
+				} else {
+					typedResult = res
+				}
+
+				if isNew {
+					// TODO: add name generation here
+					generateNameOnCreate(typedResult)
+					err = f.Tracker().Create(pa.GetResource(), typedResult, pa.GetNamespace())
+				} else if !reflect.DeepEqual(original, typedResult) {
```

Comment:
This won't work on Istio types which are protos and cannot be compared with DeepEqual

Reply:
do they always return false, or always return true, or something else?  Is there a better way to protect against apply() causing updates when nothing has changed (idempotency)?

Reply:
The new protos have these unexported cache field (state, sizeCache) that can change for otherwise identical objects. So this will return "not equal" too frequently

Reply:
I wonder how k8s does it. Maybe on the unstructured JSON?